### PR TITLE
GetGATTChildren

### PR DESCRIPTION
**Pull request by dati91**

### CODE REVIEW

File: components/script/dom/bluetooth.rs [hunk 1]
```diff
@@ -194,6 +194,57 @@ pub fn response_async<T: AsyncBluetoothListener + DomObject + 'static>(
     action_sender
 }
 
+#[allow(unrooted_must_root)]
+// https://webbluetoothcg.github.io/web-bluetooth/#getgattchildren
+pub fn get_gatt_children<T, F> (
+        attribute: &T,
+        single: bool,
+        uuidCanonicalizer: F,
```
Comment (zakorgy) [inline components/script/dom/bluetooth.rs]: use snake case

File: components/bluetooth/lib.rs [hunk 2]
```diff
@@ -678,328 +648,151 @@ impl BluetoothManager {
...
-        if !self.cached_devices.contains_key(&device_id) {
-            return drop(sender.send(Err(BluetoothError::InvalidState)));
-        }
-        let mut adapter = get_adapter_or_return_error!(self, sender);
-        let services = match uuid {
-            Some(ref id) => {
-                if !self.allowed_services.get(&device_id).map_or(false, |s| s.contains(id)) {
-                    return drop(sender.send(Err(BluetoothError::Security)))
+                // Step 6.
+                if let Some(uuid) = uuid.clone() {
```
Comment (zakorgy) [inline components/bluetooth/lib.rs]: Maybe try with `if let Some(ref uuid) = uuid`, I think this way the `clone()` can be omitted.

**Review by zakorgy — COMMENTED** [review]
Use snake case variables in ` components/script/dom/bluetooth.rs ` otherwise lgtm.

File: components/script/dom/bluetooth.rs [hunk 3]
```diff
@@ -194,6 +194,57 @@ pub fn response_async<T: AsyncBluetoothListener + DomObject + 'static>(
...
+        single: bool,
+        uuid_canonicalizer: F,
+        uuid: Option<StringOrUnsignedLong>,
+        instance_id: String,
+        connected: bool,
+        child_type: GATTType)
+        -> Rc<Promise>
+        where T: AsyncBluetoothListener + DomObject + 'static,
+              F: FnOnce(StringOrUnsignedLong) -> Fallible<UUID> {
+    let p = Promise::new(&attribute.global());
```
Comment (zakorgy) [inline components/script/dom/bluetooth.rs:210]: Just a question, if attribute is a `&T` do we need the `&` here?

**Review by zakorgy — COMMENTED** [review]
lgtm

File: components/script/dom/bluetooth.rs (hunk 3 above)
Comment (dati91) [reply components/script/dom/bluetooth.rs:210]: without it you will get the following error:
```
error[E0308]: mismatched types
   --> components/script/dom/bluetooth.rs:210:26
    |
210 |     let p = Promise::new(attribute.global());
    |                          ^^^^^^^^^^^^^^^^^^ expected reference, found struct `dom::bindings::js::Root`
    |
    = note: expected type `&dom::globalscope::GlobalScope`
    = note:    found type `dom::bindings::js::Root<dom::globalscope::GlobalScope>`
```

File: components/script/dom/bluetoothremotegattservice.rs [hunk 4]
```diff
@@ -272,49 +132,34 @@ impl AsyncBluetoothListener for BluetoothRemoteGATTService {
...
-                promise.resolve_native(promise_cx, &bt_characteristic);
-            },
-            // https://webbluetoothcg.github.io/web-bluetooth/#dom-bluetoothremotegattservice-getcharacteristics
-            // https://webbluetoothcg.github.io/web-bluetooth/#getgattchildren
-            // Step 7.
-            BluetoothResponse::GetCharacteristics(characteristics_vec) => {
+            BluetoothResponse::GetCharacteristics(characteristics_vec, single) => {
+                if single {
+                    promise.resolve_native(promise_cx,
+                                           &device.get_or_create_characteristic(&characteristics_vec[0], &self));
```
Comment (mmatyas) [inline components/script/dom/bluetoothremotegattservice.rs:140]: would this be over the max line length?

Comment (dati91) [reply components/script/dom/bluetoothremotegattservice.rs:140]: the 140. line length is 114.

**Pull request merged**